controllers/metal3.io/bmceventsubscription_controller.go

```diff
@@ -0,0 +1,260 @@
+	if err != nil {
+		return ctrl.Result{}, errors.Wrap(err, "failed to check services availability")
+	}
+	if !ready {
+		reqLogger.Info("provisioner is not ready", "RequeueAfter:", provisionerNotReadyRetryDelay)
+		return ctrl.Result{Requeue: true, RequeueAfter: provisionerNotReadyRetryDelay}, nil
+	}
+
+	httpHeaders, err := r.getHTTPHeaders(*subscription)
+
+	if err != nil {
+		reqLogger.Error(err, "failed to get http headers")
+		return ctrl.Result{RequeueAfter: subscriptionRetryDelay}, nil
+	}
+
+	if subscription.DeletionTimestamp.IsZero() {
+		// Not being deleted
+		if subscription.Status.SubscriptionID == "" {
+			if _, err := prov.AddBMCEventSubscriptionForNode(subscription, httpHeaders); err != nil {
+				return ctrl.Result{}, errors.Wrap(err, "failed to add a subscription")
+			}
+
+			// Update the subscription with the new ID
+			err := r.Status().Update(ctx, subscription)
+			if err != nil {
+				return ctrl.Result{}, errors.Wrap(err, "failed to update subscription status")
+			}
+
+		} else {
+			// Update an existing subscription by recreating it in the BMC
+
+			originalSubscription := subscription.DeepCopy()
+
+			if _, err := prov.AddBMCEventSubscriptionForNode(subscription, httpHeaders); err != nil {
+				return ctrl.Result{}, errors.Wrap(err, "failed to update subscription")
+			}
+
+			if _, err := prov.RemoveBMCEventSubscriptionForNode(*originalSubscription); err != nil {
+				return ctrl.Result{}, errors.Wrap(err, "failed to remove a subscription")
+			}
+
+			return ctrl.Result{}, nil
+		}
+	} else {
+		// being deleted
+		if subscriptionHasFinalizer(subscription) {
+			if _, err := prov.RemoveBMCEventSubscriptionForNode(*subscription); err != nil {
+				return ctrl.Result{}, errors.Wrap(err, "failed to remove a subscription")
+			}
+
+			// Remove finalizer to allow deletion
+			subscription.Finalizers = utils.FilterStringFromList(
+				subscription.Finalizers, metal3v1alpha1.BMCEventSubscriptionFinalizer)
+			reqLogger.Info("cleanup is complete, removed finalizer",
+				"remaining", subscription.Finalizers)
+			if err := r.Update(context.Background(), subscription); err != nil {
+				return ctrl.Result{}, err
+			}
+		}
+
+		// Stop reconciliation as the item is being deleted
+		return ctrl.Result{}, nil
+	}
+
+	return
+
+}
+
+func (r *BMCEventSubscriptionReconciler) getBMCSecret(request ctrl.Request, host *metal3v1alpha1.BareMetalHost) (bmcCredsSecret *corev1.Secret, err error) {
+	secretKey := host.CredentialsKey()
+	bmcCredsSecret = &corev1.Secret{}
+	err = r.Get(context.TODO(), secretKey, bmcCredsSecret)
+	if err != nil {
+		if k8serrors.IsNotFound(err) {
+			return nil, &ResolveBMCSecretRefError{message: fmt.Sprintf("The BMC secret %s does not exist", secretKey)}
+		}
+		return nil, err
+	}
+
+	return bmcCredsSecret, nil
+}
+
+func (r *BMCEventSubscriptionReconciler) getHTTPHeaders(subscription metal3v1alpha1.BMCEventSubscription) ([]map[string]string, error) {
+	headers := []map[string]string{}
```

Comment:
Why not one map? Technically, it will be a map on wire anyway.

Reply:
This is the type that gophercloud expects.

Reply:
Just a reminder, Redfish won't return any information about the headers you used to create a subscription when you retrieve information about the subscription you created 

i.e you are creating a subscription with "Content-Type: application/json", the subscription is created, if you try to retrieve information about this subscription you won't have information that it had headers

Reply:
We can introduce a new type here, and move the formatting into a map further down the stack.

Reply:
If the list of maps is what Ironic expects, then I'm fine with keeping it.